core/translate/mod.rs

```diff
@@ -357,42 +337,102 @@ fn translate_pragma(
-                let kb = cache_size_unformatted.abs() * 1024;
-                kb / 512 // assume 512 page size for now
-            } else {
-                value
-            } as usize;
-            if cache_size < MIN_PAGE_CACHE_SIZE {
-                // update both in memory and stored disk value
-                cache_size = MIN_PAGE_CACHE_SIZE;
-                cache_size_unformatted = MIN_PAGE_CACHE_SIZE as i64;
-            }
-
-            // update in-memory header
-            header.borrow_mut().default_cache_size = cache_size_unformatted
-                .try_into()
-                .unwrap_or_else(|_| panic!("invalid value, too big for a i32 {}", value));
+fn update_pragma(
+    name: &str,
+    value: ast::Expr,
+    header: Rc<RefCell<DatabaseHeader>>,
+    pager: Rc<Pager>,
+    program: &mut ProgramBuilder,
+) {
+    let pragma = PragmaName::from_str(name).expect("provided pragma not valid");
+    match pragma {
+        PragmaName::CacheSize => {
+            let cache_size = match value {
+                ast::Expr::Literal(ast::Literal::Numeric(numeric_value)) => {
+                    numeric_value.parse::<i64>().unwrap()
+                }
+                ast::Expr::Unary(ast::UnaryOperator::Negative, expr) => match *expr {
+                    ast::Expr::Literal(ast::Literal::Numeric(numeric_value)) => {
+                        -numeric_value.parse::<i64>().unwrap()
+                    }
+                    _ => 0,
+                },
+                _ => 0,
+            };
 
-            // update in disk
-            let header_copy = header.borrow().clone();
-            pager.write_database_header(&header_copy);
+            update_cache_size(cache_size, header, pager);
+        }
+        PragmaName::JournalMode => {
+            query_pragma("journal_mode", header, program);
+        }
+        _ => todo!("pragma `{name}`"),
+    }
+}
 
-            // update cache size
-            pager.change_page_cache_size(cache_size);
+fn query_pragma(
+    name: &str,
+    database_header: Rc<RefCell<DatabaseHeader>>,
+    program: &mut ProgramBuilder,
+) {
+    let pragma = PragmaName::from_str(name).expect("provided pragma not valid");
+    let register = program.alloc_register();
+    match pragma {
+        PragmaName::CacheSize => {
+            program.emit_insn(Insn::Integer {
+                value: database_header.borrow().default_cache_size.into(),
+                dest: register,
+            });
         }
-        _ => todo!(),
+        PragmaName::JournalMode => {
+            program.emit_insn(Insn::String8 {
+                value: "wal".into(),
+                dest: register,
+            });
+        }
+        _ => {
+            todo!("pragma `{name}`");
+        }
+    }
+
+    let next_register = program.next_free_register();
+    program.emit_insn(Insn::ResultRow {
+        start_reg: register,
+        count: next_register - register,
```

Comment:
I think it should be safe to assume it will be 1 for every case. This arithmetic might be confusing if more registers are allocated in between that could be unrelated.